File: SmartBill/src/main/java/com/BCSTech/SmartBill/company/model/Company.java
package com.BCSTech.SmartBill.company.model;

import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "companies")
public class Company {

    @Id
    private String id;

    private String name;

    @Indexed(unique = true, sparse = true)
    private String gstNumber;

    private String panNumber;

    private String address;

    private String city;

    private String state;

    private String pincode;

    private String country;

    private String email;

    private String mobile;

    private String logoUrl;

    private String signatureUrl;

    // Default currency for invoices — INR for Indian businesses
    @Builder.Default
    private String defaultCurrency = "INR";

    // Invoice number prefix e.g. "INV-"
    @Builder.Default
    private String invoicePrefix = "INV-";

    // Starting invoice number — incremented on each invoice creation
    @Builder.Default
    private int invoiceSequence = 1;

    // Financial year start month (4 = April for Indian FY)
    @Builder.Default
    private int financialYearStartMonth = 4;

    @Builder.Default
    private SubscriptionPlan subscriptionPlan = SubscriptionPlan.FREE;

    // Owner/admin userId who created this company
    private String createdByUserId;

    @Builder.Default
    private boolean active = true;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;
}

File: frontend/src/api/company.ts
import { apiClient } from "./client";
import type { ApiResponse } from "../types/api";

export type SubscriptionPlan = "FREE" | "BASIC" | "PRO" | "ENTERPRISE";

// Mirrors CompanyDTOs.CompanyResponse
export interface CompanyResponse {
  id: string;
  name: string;
  gstNumber?: string;
  panNumber?: string;
  address?: string;
  city?: string;
  state?: string;
  pincode?: string;
  country?: string;
  email?: string;
  mobile?: string;
  logoUrl?: string;
  signatureUrl?: string;
  defaultCurrency?: string;
  invoicePrefix?: string;
  invoiceSequence: number;
  financialYearStartMonth: number;
  subscriptionPlan: SubscriptionPlan;
}

// Mirrors CompanyDTOs.UpdateRequest
export interface CompanyUpdateRequest {
  name: string;
  gstNumber?: string;
  panNumber?: string;
  address?: string;
  city?: string;
  state?: string;
  pincode?: string;
  country?: string;
  email?: string;
  mobile?: string;
  logoUrl?: string;
  signatureUrl?: string;
  defaultCurrency?: string;
  invoicePrefix?: string;
  financialYearStartMonth?: number;
}

async function unwrap<T>(promise: Promise<{ data: ApiResponse<T> }>): Promise<T> {
  const { data } = await promise;
  if (data.data === undefined) throw new Error(data.message ?? "Empty response from server");
  return data.data;
}

export const companyApi = {
  get: () => unwrap<CompanyResponse>(apiClient.get("/api/company")),

  update: (body: CompanyUpdateRequest) =>
    unwrap<CompanyResponse>(apiClient.put("/api/company", body)),
};
